test/e2e/network/service.go

```diff
@@ -2027,6 +2027,225 @@ var _ = common.SIGDescribe("Services", func() {
 		}
 	})
 
+	ginkgo.It("should respect internalTrafficPolicy=Local when there are endpoints [Feature:ServiceInternalTrafficPolicy]", func() {
+		// windows kube-proxy does not support this feature yet
+		e2eskipper.SkipIfNodeOSDistroIs("windows")
+
+		// This behavior is not supported if Kube-proxy is in "userspace" mode.
+		// So we check the kube-proxy mode and skip this test if that's the case.
+		if proxyMode, err := proxyMode(f); err == nil {
+			if proxyMode == "userspace" {
+				e2eskipper.Skipf("The test doesn't work with kube-proxy in userspace mode")
+			}
+		} else {
+			framework.Logf("Couldn't detect KubeProxy mode - test failure may be expected: %v", err)
+		}
+
+		serviceName := "svc-itp"
+		ns := f.Namespace.Name
+		servicePort := 80
+
+		ginkgo.By("creating a TCP service " + serviceName + " with type=ClusterIP and internalTrafficPolicy=Local in namespace " + ns)
+		jig := e2eservice.NewTestJig(cs, ns, serviceName)
+		jig.ExternalIPs = true
+
+		local := v1.ServiceInternalTrafficPolicyLocal
+		svc, err := jig.CreateTCPService(func(svc *v1.Service) {
+			svc.Spec.Ports = []v1.ServicePort{
+				{Port: 80, Name: "http", Protocol: v1.ProtocolTCP, TargetPort: intstr.FromInt(80)},
+			}
+			svc.Spec.InternalTrafficPolicy = &local
+		})
+		framework.ExpectNoError(err)
+		defer func() {
+			framework.Logf("Cleaning up the webserver service")
+			err := cs.CoreV1().Services(ns).Delete(context.TODO(), serviceName, metav1.DeleteOptions{})
+			framework.ExpectNoError(err, "failed to delete service: %s in namespace: %s", serviceName, ns)
+		}()
+		serviceIP := svc.Spec.ClusterIP
+		framework.Logf("internal-traffic-policy-test cluster ip: %s", serviceIP)
+		ginkgo.By("Picking 2 Nodes to test whether internal traffic policy is respected")
+		nodes, err := e2enode.GetBoundedReadySchedulableNodes(cs, 2)
+		framework.ExpectNoError(err)
+		nodeCounts := len(nodes.Items)
+		if nodeCounts < 2 {
+			e2eskipper.Skipf("The test requires at least two ready nodes on %s, but found %v", framework.TestContext.Provider, nodeCounts)
+		}
+
+		ginkgo.By("Creating 2 webserver pods to be part of the TCP service")
+		webserverPod0 := e2epod.NewAgnhostPod(ns, "echo-hostname-0", nil, nil, nil, "netexec", "--http-port", strconv.Itoa(servicePort))
+		webserverPod0.Labels = jig.Labels
+		webserverPod0.Spec.NodeName = nodes.Items[0].Name
```

Comment:
we shouldn't set the NodeName directly, it bypass the scheduler and create flakes
xref: https://github.com/kubernetes/kubernetes/pull/101184

Reply:
hmmm... but setting NodeName explicitly simplifies the test a lot, because we know deterministically which pods should work for local traffic and which ones will fail.

Reply:
I guess we can use nodeSelector and hostname label?

Reply:
yeah, sorry, I should be more explicit, check the reference, 

```
-		clientPod.Spec.NodeName = clientNodeInfo.name
+		nodeSelection := e2epod.NodeSelection{Name: clientNodeInfo.name}
+		e2epod.SetNodeSelection(&clientPod.Spec, nodeSelection)
```

Reply:
thanks, updated to use these